Hole average stats now reflect recent round performance.

makinde opened the pull request:
This is accomplished by having the data generator record an additional stat per player/hole that only reflects the performance in the last X holes. The client then has access to all time data and this recent metric. The data generation change is so slight, it should not add any additional maintenance cost.

This change allows us see how folks are doing in their latest rounds, instead of displayed "career averages". This also allows us to show arrows signifying if players are trending up or down on given holes when compared to their lifetime average.

There are minimal changes here to the data generation. I agree with most that the majority/all of the data generation should just happen on the fly in the web page instead of statically generating it all upfront. That would be a great change for someone to make separate of this change.

fzamore commented on data/generateData.js:
```diff
@@ -60,14 +61,17 @@ async function parseSheet(sheet) {
     const avgScore = getAvg(cumScore, roundsPlayed);
 
+    // only include first X rounds for a given player in their stats
+    if (roundsPlayed > maxRoundsPerPlayer) return;
```
This isn't 'nam!

```suggestion
    if (roundsPlayed > maxRoundsPerPlayer) {
      return;
    }
```

makinde replied:
There are lint rules in this repo and this style passed those rules. Wont fix

fzamore replied:
sadpanda

Guess I've gotta add a lint rule to prevent abominations like this, then!

ericgio replied:
https://eslint.org/docs/rules/curly

fzamore commented on data/generateData.js:
```diff
@@ -177,8 +181,11 @@ function getTopRoundsByScore(roundsByScore) {
-  const statsByPlayer = Object.values(rounds.reduce(getStatsByPlayer, {}));
+function getStats(rounds, maxRoundsPerPlayer) {
+  const boundStatsByPlayer = getStatsByPlayer.bind(null, maxRoundsPerPlayer);
+  // Recent rounds first so the maxRoundsPerPlayer will the right ones
```
```suggestion
  // Ensure recent rounds come first so the maxRoundsPerPlayer will choose the right ones
```

fzamore commented on data/generateData.js:
```diff
@@ -177,8 +181,11 @@ function getTopRoundsByScore(roundsByScore) {
-function getStats(rounds) {
-  const statsByPlayer = Object.values(rounds.reduce(getStatsByPlayer, {}));
+function getStats(rounds, maxRoundsPerPlayer) {
+  const boundStatsByPlayer = getStatsByPlayer.bind(null, maxRoundsPerPlayer);
+  // Recent rounds first so the maxRoundsPerPlayer will the right ones
+  rounds.sort((a, b) => new Date(b.date) - new Date(a.date));
+  const statsByPlayer = Object.values(rounds.reduce(boundStatsByPlayer, {}));
```
This `reduce` stuff still confuses me, but it's certainly in-line with what the rest of this file does, so if it works, I'm fine with it.

makinde replied:
I'm also not a fan of it. I'm trying to make a minor tweak, not do heart surgery.

fzamore commented on pages/stats.js:
```diff
@@ -82,15 +83,21 @@ const SortableHeader = ({
+  const [selectedSortKey, setSelectedSortKey] = useState('avgScore');
+  const [statsRange, setStatsRange] = useState('recent');
+
+  let relevantStats;
+  if (statsRange === 'recent') relevantStats = recentStats;
+  if (statsRange === 'allTime') relevantStats = stats;
+
```
Constants for `'recent'` and `'allTime'`? (Avoid hard-coding twice.)

Also, I'd just use a `const` + ternary to avoid the braces-less if-statements:
```suggestion
  const relevantStats = (statsRange === 'recent') ? recentStats : stats;
```

fzamore commented on pages/stats.js:
```diff
@@ -82,15 +83,21 @@ const SortableHeader = ({
     statsByPlayer,
     MIN_ROUNDS,
-  } = stats;
-  const [order, setOrder] = useState('desc');
-  const [selectedSortKey, setSelectedSortKey] = useState('avgScore');
```
I'm confused - why are you deleting these? They're still being used, right?

makinde replied:
They are just being by moved a few lines up.

fzamore reviewed (approved):
You asked for it - here's an annoying Zamore review!

ericgio commented on data/generateData.js:
```diff
@@ -177,8 +181,11 @@ function getTopRoundsByScore(roundsByScore) {
-function getStats(rounds) {
-  const statsByPlayer = Object.values(rounds.reduce(getStatsByPlayer, {}));
+function getStats(rounds, maxRoundsPerPlayer) {
+  const boundStatsByPlayer = getStatsByPlayer.bind(null, maxRoundsPerPlayer);
```
Let's not do this. It's super-confusing to change the standard signature for a reduce function.

makinde replied:
Okay, I’ll change it to a function that returns a function with the standard sig.

ericgio commented on data/generateData.js:
```diff
@@ -177,8 +181,11 @@ function getTopRoundsByScore(roundsByScore) {
 
-function getStats(rounds) {
-  const statsByPlayer = Object.values(rounds.reduce(getStatsByPlayer, {}));
+function getStats(rounds, maxRoundsPerPlayer) {
```
Add a default value here so you don't need to pass a value everywhere. Also, `MAX_SAFE_INTEGER` seems a bit excessive. At our current rate, we'd barely break 1000 rounds if we played consistently for the next 40 years.

```suggestion
function getStats(rounds, maxRoundsPerPlayer = Number.MAX_SAFE_INTEGER) {
```

makinde replied:
I personally think the default value is confusing. It means different call sites (only two) have different Params and means u have to look in more places to see what’s going on. But I’ll change it.

makinde replied:
Also, defensive programming! Otherwise 1,000 becomes a magic number that is pulled out of thin air.

fzamore replied:
FWIW, I prefer the default value too. Don't feel strongly about 1000 vs. MAX_INT.

makinde replied:
Removed need for this in latest rev

ericgio reviewed (commented):
I think the static data generation makes much less sense with a change like this. That approach assumed the data would be... static. If we want to filter and update the displayed data, it probably makes more sense to do (at least some of) these transformations on the client.

makinde commented:
Okay, I believe I've addressed the comments. The new change to the data generation code is now much simpler (recording one extra stat per hole). And the stuff in the stats page is also simpler (and IMHO, much better functionality wise). The new functionality only changes the Avg score by hole table (the middle one) to make it reflect scores in each player's most recent 8 rounds (4 months for those of us playing often). It shows a trending arrow next to the score if your avg in that time is significantly lower/higher than your lifetime score. The lifetime score is still displayed in a tooltip, so no information is lost. IMHO, recently data is so much more interesting that the relatively stale, hard to move, lifetime average.

I totally get that so much more can be done here with graphs and other high flying things, but I'm hoping that DONE can be better than perfect here. Merge pls?

makinde commented:
![Uploading Screen Shot 2020-09-28 at 4.29.21 PM.png…]()

fzamore commented:
Dang, this change would give me a sub-4 average on Hole 2? Man, now we *have* to take this change.

fzamore commented on data/generateData.js:
```diff
@@ -75,10 +76,15 @@ function getStatsByPlayer(acc, { players }, index, rounds) {
       const cumHoleScore = idx(scoresByHole[hole], 'cumHoleScore', 0) + score;
+      const avgHoleScore = getAvg(cumHoleScore, roundsPlayed);
+      const recentAvgHoleScore = roundsPlayed > RECENT_ROUND_COUNT ?
+        idx(scoresByHole[hole], 'recentAvgHoleScore', 0) :
```
I cannot figure out how the value for `recentAvgHoleScore` gets populated. I guess I'll merge it and play around with it.

makinde replied:
It's the oddity of the accumulator stuff. ScoresByHole is taken from the `acc`, which is basically the "result so far". So the `idx` is saying, ~"try to get this value from the result so far. if it's not there already, use this default value".

fzamore replied:
Okay, I get it now. The scores are sorted in *reverse* chronological order (i.e., most recent first). So, starting from the 9th round on (when `roundsPlayed > RECENT_ROUND_COUNT`), `'recentAvgHoleScore'` never changes.

fzamore reviewed (approved):
Cool. I will try to merge and then manually make followups. We gotta get you as a contributor.

makinde commented on data/generateData.js:
```diff
@@ -18,6 +18,7 @@ const CELL_RANGE = 'A3:J';
 const PAR = 27;
 const TOP_ROUNDS = 10;
 const MIN_ROUNDS = 3;
+const RECENT_ROUND_COUNT = 8;
```
Realized that with biweekly games, 10 rounds ~= 5 months, which seemed like a really long time. 3 months felt more reasonable, but I was worried that data would swing around too much. Down for different numbers here if anyone cares.

makinde commented on pages/stats.js:
```diff
@@ -166,7 +176,7 @@ const StatsPage = ({ stats }) => {
       </Table>
 
-      <h3>Average Score By Hole (minimum {MIN_ROUNDS} rounds)</h3>
+      <h3>Recent Average Scores By Hole (minimum {MIN_ROUNDS} rounds)</h3>
```
Sure, or maybe even hide the details behind a tooltip or something. Both titles are really long.
